---
title: 'Preprocessing the ATAC-seq for both ES/EB and MEF
output: html_document
date: '04/23/2019'
author: 'Wuming Gong'
---

```{r, message = FALSE}
library(BSgenome.Mmusculus.UCSC.mm10)
library(dplyr)
library(futile.logger)
library(BiocParallel)
register(MulticoreParam(8))
devtools::load_all('../ngsmsi')
options(width = as.integer(system('tput cols', intern = TRUE)))
```

### Process the ATAC-seq for ES/EB differentiation and MEF reprogramming induced by Etv2
* Note that all reads have been concatenated across lanes within each flow cell.  
However, to achieve full read depth, your informatics team will need to concatenate reads across flow cells.
```{r}
root_dir <- sprintf('%s/etv2_atac', Sys.getenv('TMPDIR'))
if (!file.exists(root_dir)) dir.create(root_dir)
run1 <- '/home/garrydj/data_release/umgc/nextseq/190227_NB551164_0120_AHJYKYBGX9/Garry_Project_066_068_Run1'
run2 <- '/home/garrydj/data_release/umgc/nextseq/190227_NB551498_0015_AHK23NBGX9/Garry_Project_066_068_Run2'
files1 <- list.files(run1, 'fastq.gz', full.names = FALSE)
files2 <- list.files(run2, 'fastq.gz', full.names = FALSE)
all(files1 == files2) # this should be TRUE
res <- bplapply(1:length(files1), function(i){
	command <- sprintf('cat %s/%s %s/%s > %s/%s', run1, files1[i], run2, files1[i], root_dir, files1[i])
	flog.info(sprintf('%d/%d | %s', i, length(files1), command))
	system(command)
})
```

### Move the fastq files to the temdir for alignment
```{r}
study_accession <- 'etv2_atac'
dataset <- 'dataset=Etv2ATAC_version=20190228b'
root <- Sys.getenv('TMPDIR')
seqtype <- 'ATAC'
genome <- 'mm10'
data.dir <- sprintf('%s/%s', root, study_accession)
fastq.files <- list.files(data.dir, 'fastq.gz', full.names = TRUE)
run_accession <- gsub(sprintf('%s/(.+?)_R[1|2].+$', data.dir), '\\1', fastq.files)
pairend <- as.numeric(gsub(sprintf('%s/.+?_R([1|2]).+$', data.dir), '\\1', fastq.files))
res <- sapply(1:length(fastq.files), function(i) dir.create(sra_run_dir(run_accession[i]), recursive = TRUE))  # create run dir
target.files <- sprintf('%s/%s_%d.fastq.gz', sra_run_dir(run_accession), run_accession, pairend)
res <- bplapply(1:length(fastq.files), function(i){ 
	command <- sprintf('cp %s %s', fastq.files[i], target.files[i])
	cat(sprintf('[%s] %s\n', Sys.time(), command))
	system(command)
})  # move fastq files into tmpdir
```

### Prepare a data frame for each replicate
```{r}
run2group <- c(
	'Dox_D1-1_S3' = 'MEF_Dox_D1_rep1',
	'Dox_D1-2_S4' =	'MEF_Dox_D1_rep2',
	'Dox_D2-1_S5' =	'MEF_Dox_D2_rep1',
	'Dox_D2-2_S6' =	'MEF_Dox_D2_rep2',
	'Dox_D7-1_Bulk_S7' = 'MEF_Dox_D7_rep1',
	'Dox_D7-1_Sorted_S9' = 'MEF_Dox_D7_Flk1pos_rep1',
	'Dox_D7-2_Bulk_S8' = 'MEF_Dox_D7_rep2',
	'Dox_D7-2_Sorted_S10' =	'MEF_Dox_D7_Flk1pos_rep2',
	'EB_D25_Dox12hBulk1_S13' = 'EB_Dox_D25_rep1',
	'EB_D25_Dox12hBulk2_S14' = 'EB_Dox_D25_rep2',
	'EB_D25_Dox12hSorted1_S15' = 'EB_Dox_D25_Flk1pos_rep1',
	'EB_D25_Dox12hSorted2_S16' = 'EB_Dox_D25_Flk1pos_rep2',
	'EB_D25_NoDox1_S11'	= 'EB_NoDox_D25_rep1',
	'EB_D25_NoDox2_S12'	= 'EB_NoDox_D25_rep2',
	'No_Dox_1_S1'	= 'MEF_NoDox_rep1',
	'No_Dox_2_S2' =	'MEF_NoDox_rep2'
)
group <- run2group[unique(run_accession)]
d <- data.frame(study_accession = study_accession, run = unique(run_accession), seqtype = seqtype, genome = genome, group = group)
write.table(d, sprintf('../datasets/%s.tsv', dataset), sep = '\t', quote = FALSE, row.names = FALSE)
```

# Calling ATAC-seq peaks for EB Etv2 induction at D2.5 and MEF Etv2 induction data

# Calling ATAC-seq peaks for Naoko's ATAC-seq on D3 EB with and without Etv2 induction
```{r, eval = FALSE}
dataset <- 'dataset=Naoko_version=20170628a'; genome <- 'mm10'
d <- read_dataset(dataset)
d <- d %>% mutate(bam_file = sra.dedup.bam.file(run))

bname <- 'EB_Dox_D3'; treatment <- c('d3_EB_Dox_1_S3', 'd3_EB_Dox_2_S4')
#bname <- 'EB_NoDox_D3'; treatment <- c('d3_EB_no_dox_1_S1', 'd3_EB_no_dox_2_S2')

base.name <- sprintf('%s/%s', dataset_dir(dataset), bname); treatment_files <- d[d[, 'run'] %in% treatment, 'bam_file']
macs2.callpeak(treatment_files, base.name, format = 'BAMPE', genome = genome, broad = FALSE, qvalue.cutoff = 0.05, fold.change = FALSE, update = TRUE, call.summits = TRUE, shift = -100, extsize = 200)
```

# Calling ATAC-seq peaks for Naoko's ATAC-seq on MEF and D2 EB with and without Etv2 induction
```{r, eval = FALSE}
dataset <- 'dataset=Naoko_version=20170302a'; genome <- 'mm10'
d <- read_dataset(dataset)
d <- d %>% mutate(bam_file = sra.dedup.bam.file(run))

#bname <- 'EB_Dox_D3_Flk1pos'; treatment <- c('D3EB_F_P_6h_Dox-1_S12', 'D3EB_F_P_6h_Dox-2_S13')
#bname <- 'EB_NoDox_D3_Flk1pos'; treatment <- c('D3EB_F_P_no_Dox-1_S10', 'D3EB_F_P_no_Dox-2_S11')
#bname <- 'EB_Dox_D0'; treatment <- c('ES_6h_Dox-1_S8', 'ES_6h_Dox-2_S9')
bname <- 'EB_NoDox_D0'; treatment <- c('ES_no_Dox-1_S6', 'ES_no_Dox-2_S7')

base.name <- sprintf('%s/%s', dataset_dir(dataset), bname); treatment_files <- d[d[, 'run'] %in% treatment, 'bam_file']
macs2.callpeak(treatment_files, base.name, format = 'BAMPE', genome = genome, broad = FALSE, qvalue.cutoff = 0.05, fold.change = FALSE, update = TRUE, call.summits = TRUE, shift = -100, extsize = 200)
```

# Merging sample-level BAM files (dedup.bam) into group-level BAM files
```{r}
dataset <- 'dataset=Naoko_version=20170302a'; bname <- 'EB_Dox_D0'; runs <- c('ES_6h_Dox-1_S8', 'ES_6h_Dox-2_S9')
dataset <- 'dataset=Naoko_version=20170302a'; bname <- 'EB_NoDox_D0'; runs <- c('ES_no_Dox-1_S6', 'ES_no_Dox-2_S7')
dataset <- 'dataset=Naoko_version=20170628a'; bname <- 'EB_Dox_D3'; runs <- c('d3_EB_Dox_1_S3', 'd3_EB_Dox_2_S4')
dataset <- 'dataset=Naoko_version=20170628a'; bname <- 'EB_NoDox_D3'; runs <- c('d3_EB_no_dox_1_S1', 'd3_EB_no_dox_2_S2')
dataset <- 'dataset=Naoko_version=20170302a'; bname <- 'EB_Dox_D3_Flk1pos'; runs <- c('D3EB_F_P_6h_Dox-1_S12', 'D3EB_F_P_6h_Dox-2_S13')
dataset <- 'dataset=Naoko_version=20170302a'; bname <- 'EB_NoDox_D3_Flk1pos'; runs <- c('D3EB_F_P_no_Dox-1_S10', 'D3EB_F_P_no_Dox-2_S11')

d <- read_dataset(dataset)
d <- d %>% mutate(bam_file = sra.dedup.bam.file(run))
input_bam_files <- d[d$run %in% runs, 'bam_file']
output_bam_file <- sprintf('%s/%s.bam', dataset_dir(dataset), bname)

merge.bam.files(treatment_files, output_bam_file)
```

# Define a union set of ATAC-seq summits and save as a bed file:
```{r}
bed_files <- c(

	'EB_Dox_D0' = '/panfs/roc/scratch/gongx030/datasets/dataset=Naoko_version=20170302a/EB_Dox_D0_summits.bed',
	'EB_NoDox_D0' = '/panfs/roc/scratch/gongx030/datasets/dataset=Naoko_version=20170302a/EB_NoDox_D0_summits.bed',
	'EB_Dox_D3_Flk1pos' = '/panfs/roc/scratch/gongx030/datasets/dataset=Naoko_version=20170302a/EB_Dox_D3_Flk1pos_summits.bed',
	'EB_NoDox_D3_Flk1pos' = '/panfs/roc/scratch/gongx030/datasets/dataset=Naoko_version=20170302a/EB_NoDox_D3_Flk1pos_summits.bed',

	'EB_Dox_D3' = '/panfs/roc/scratch/gongx030/datasets/dataset=Naoko_version=20170628a/EB_Dox_D3_summits.bed',
	'EB_NoDox_D3' = '/panfs/roc/scratch/gongx030/datasets/dataset=Naoko_version=20170628a/EB_NoDox_D3_summits.bed',

	'MEF_NoDox' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_NoDox_summits.bed',
	'MEF_Dox_D1' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D1_summits.bed',
	'MEF_Dox_D2' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D2_summits.bed',
  'MEF_Dox_D7' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D7_summits.bed',
  'MEF_Dox_D7_Flk1pos' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D7_Flk1pos_summits.bed',

  'EB_Dox_D25' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/EB_Dox_D25_summits.bed',
  'EB_Dox_D25_Flk1pos' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/EB_Dox_D25_Flk1pos_summits.bed',
  'EB_NoDox_D25' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/EB_NoDox_D25_summits.bed'
)

gr_list <- lapply(1:length(bed_files), function(i){
  x <- macs2.read_summits(bed_files[i])
	x$source <- names(bed_files)[i]
  x
})
gr <- Reduce('c', gr_list)
gr <- resize(gr, width = 200, fix = 'center')
gr0 <- reduce(gr) # a reduced set of Etv2 peak
mm <- as.matrix(findOverlaps(gr, gr0))
sp <- split(mm[, 1], list(mm[, 2]))
j <- unlist(mclapply(sp, function(i) i[which.max(gr$score[i])], mc.cores = 4))
gr <- gr[j]
G <- do.call('cbind', lapply(1:length(bed_files), function(i) gr %over% gr_list[[i]]))
gr$group <- G # group assignment of the union set to each source
colnames(gr$group) <- names(bed_files)
saveRDS(gr, '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/all_ATAC_seq_peaks.rds')
s3_saveRDS(gr, 'datasets/dataset=Etv2ATAC_version=20190228a/all_ATAC_seq_peaks.rds')
bed_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/all_ATAC_seq_peaks.bed'
write.table(as.data.frame(gr), bed_file, sep = '\t', quote = FALSE, row.names = FALSE, col.names = FALSE)
```

# Generating a read counts matrix for each ATAC-seq sample, over a union peak set across all ATAC-seq samples
These data will be used to perform a sample level chromVAR analysis
```{r}
bed_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/all_ATAC_seq_peaks.bed'
peaks <- macs2.read_summits(bed_file)
peaks <- add.seqinfo(peaks, 'mm10')

dataset <- 'dataset=Etv2ATAC_version=20190228b'
d <- read_dataset(dataset)
d <- d %>% mutate(bw_file = sprintf('%s/%s_treat_pileup.bw', dataset_dir(dataset), group))
d <- d %>% mutate(group = gsub('_rep\\d$', '', group))

dataset2 <- 'dataset=Naoko_version=20170302a'
d2 <- read_dataset(dataset2)
g2 <- c(
	'd0_EB_dox' = 'jB_Dox_D0',
	'd0_EB_no_dox' = 'EB_NoDox_D0',
	'd3_EB_dox' = 'EB_Dox_D3_Flk1pos',
	'd3_EB_no_dox' = 'EB_NoDox_D3_Flk1pos'
)
d2 <- d2 %>% mutate(
	bw_file = sprintf('%s/%s_treat_pileup.bw', dataset_dir(dataset2), run),
	group = g2[group]
)

dataset3 <- 'dataset=Naoko_version=20170628a'
d3 <- read_dataset(dataset3)
g3 <- c(
	'd2_EB_dox' = 'EB_Dox_D3',
	'd2_EB_no_dox' = 'EB_NoDox_D3'
)
d3 <- d3 %>% mutate(
	bw_file = sprintf('%s/%s_treat_pileup.bw', dataset_dir(dataset3), run),
	group = g3[group]
)

d <- rbind(d, d2, d3)
file.exists(d$bw_file)	# make sure the BAM files exist

X <- do.call('cbind', mclapply(d$bw_file, function(bw_file){
  flog.info(sprintf('reading %s', bw_file))
#	ga <- rtracklayer::import(bw_file, format = 'BigWig', which = reduce(peaks))
	ga <- rtracklayer::import(bw_file, format = 'BigWig')#, which = reduce(peaks))
	ga <- add.seqinfo(ga, 'mm10')
  cvg <- coverage(ga, weight = as.numeric(mcols(ga)$score))
  sum(cvg[peaks])
}, mc.cores = 4))

se <- SummarizedExperiment(assays = SimpleList(counts = X), rowRanges = peaks, colData = d)
s3_saveRDS(se, sprintf('gongx030/datasets/%s/ATAC_peaks_Etv2_reprogramming.rds', dataset))

```

# HINT analysis for ATAC-seq footprinting
These are run in batch by jobs 20191215*


# Merging the nucleoATAC results
These are run in batch by jobs 20191216*
```{r}

dataset <- 'dataset=Etv2ATAC_version=20190228a'; bname <- 'MEF_NoDox'
dataset <- 'dataset=Etv2ATAC_version=20190228a'; bname <- 'MEF_Dox_D1'

base_name <- sprintf('%s/%s', dataset_dir(dataset), bname)
devtools::load_all('packages/compbio'); nucleoatac_merge(base_name, job = '20191215', genome = BSgenome.Mmusculus.UCSC.mm10)

```

