---
title: 'Preprocessing the ChIP-seq for both ES/EB and MEF
output: html_document
date: '12/13/2019'
author: 'Wuming Gong'
---

```{r}
devtools::load_all('packages/compbio')
library(BSgenome.Mmusculus.UCSC.mm10)
library(dplyr)
library(parallel)
```

Calling Peaks for Etv2, Brg1 and H3K27ac ChIP-seq
```{r}
dataset <- 'dataset=Etv2PioneerChIPseq_version=20191203a'
genome <- 'mm10'
d <- read.table('https://docs.google.com/spreadsheets/d/e/2PACX-1vQvPeb6eRcvT7kY0ymiVim8ScprZhBW7DIoHlVqzIAc_cDuVmMqR777jUZU0XuTVxSzDYw0BTbrVMTZ/pub?output=tsv', sep = '\t', header = TRUE)
d <- d[d$status == 'good', ]

#treatment <- 'Etv2'; control <- 'input'; celltype <- 'MEF'; stage <- 'd1'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Etv2'; control <- 'input'; celltype <- 'MEF'; stage <- 'd2'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Etv2'; control <- 'input'; celltype <- 'MEF'; stage <- 'd7'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Etv2'; control <- 'input'; celltype <- 'EB'; stage <- '3h'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Etv2'; control <- 'input'; celltype <- 'EB'; stage <- '12h'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'MEF'; stage <- 'd0'; dox_status <- 'NoDox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'MEF'; stage <- 'd1'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'MEF'; stage <- 'd2'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'MEF'; stage <- 'd7'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'EB'; stage <- '3h'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'EB'; stage <- '3h'; dox_status <- 'NoDox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'EB'; stage <- '12h'; dox_status <- 'Dox'; broad <- FALSE; call_summits <- TRUE
#treatment <- 'Brg1'; control <- 'input'; celltype <- 'EB'; stage <- '12h'; dox_status <- 'NoDox'; broad <- FALSE; call_summits <- TRUE
treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'MEF'; stage <- 'd0'; dox_status <- 'NoDox'; broad <- TRUE; call_summits <- FALSE
#treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'MEF'; stage <- 'd1'; dox_status <- 'Dox'; broad <- TRUE; call_summits <- FALSE
#treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'MEF'; stage <- 'd2'; dox_status <- 'Dox'; broad <- TRUE; call_summits <- FALSE
#treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'MEF'; stage <- 'd7'; dox_status <- 'Dox'; broad <- TRUE; call_summits <- FALSE
#treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'EB'; stage <- '3h'; dox_status <- 'Dox'; broad <- TRUE; call_summits <- FALSE
#treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'EB'; stage <- '12h'; dox_status <- 'Dox'; broad <- TRUE; call_summits <- FALSE
#treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'EB'; stage <- '3h'; dox_status <- 'NoDox'; broad <- TRUE; call_summits <- FALSE
#treatment <- 'H3K27ac'; control <- 'input'; celltype <- 'EB'; stage <- '12h'; dox_status <- 'NoDox'; broad <- TRUE; call_summits <- FALSE

bname <- sprintf('%s_%s_%s_%s', celltype, dox_status, stage, treatment)
treatment_files <- d %>% filter(cell == celltype & time == stage & dox == dox_status & seqtype == treatment) %>% select(bam_file) %>% unlist()
control_files <- d %>% filter(cell == celltype & time == stage & dox == dox_status & seqtype == 'input') %>% select(bam_file) %>% unlist()
base.name <- sprintf('%s/%s', dataset_dir(dataset), bname)

macs2.callpeak(treatment_files, base.name, control_files, format = 'BAM', genome = genome, broad = broad, qvalue.cutoff = 0.05, fold.change = TRUE, update = TRUE, call.summits = call_summits)
```

