---
title: "chromVAR analysis of combined MEF and ES/EB ATAC-seq"
author: "Wuming Gong"
date: "12/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5, fig.height =5, dpi = 300)
```

```{r, message = FALSE}
library(chromVARmotifs) # https://github.com/GreenleafLab/chromVARmotifs
library(chromVAR)
library(BSgenome.Mmusculus.UCSC.mm10)
library(motifmatchr)
library(BiocParallel)
library(SummarizedExperiment)
library(gplots)
library(circlize)
library(ComplexHeatmap)
register(MulticoreParam(4)) # Use 8 cores
```

Read the ATAC-seq read counts data for each sample and count the motifs at each interval
```{r}
se <- readRDS(gzcon(url('https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/ATAC_peaks_Etv2_reprogramming.rds')))
data('homer_pwms')
se <- addGCBias(se, genome = BSgenome.Mmusculus.UCSC.mm10)
motif_ix <- matchMotifs(homer_pwms, se, genome = 'mm10')
dev <- computeDeviations(object = se, annotations = motif_ix)
v <- computeVariability(dev)
```


```{r, fig.width = 5, fig.height =5, dpi = 300}
m <- v$p_value_adj < 0.05
s <- prcomp(assays(dev)$z[m, ])
eigs <- s$sdev^2

colData(se)$cell <- gsub('(.+?)_.+', '\\1', colData(se)$group)
colData(se)$dox <- rep("Dox", ncol(se))
colData(se)$dox[grep('NoDox', colData(se)$group)] <- 'NoDox'
MEF_cols <- colorpanel(4, low = 'black', high = 'green')
EB_cols <- colorpanel(3, low = 'gray', high = 'red')
group2col <- c(
  'MEF_NoDox' = MEF_cols[1],
  'MEF_Dox_D1' = MEF_cols[2],
  'MEF_Dox_D2' = MEF_cols[3],
  'MEF_Dox_D7' = MEF_cols[4],
  'MEF_Dox_D7_Flk1pos' = MEF_cols[4],
  
  'EB_NoDox_D0' = EB_cols[1],
  'EB_NoDox_D25' = EB_cols[2],
  'EB_NoDox_D3' = EB_cols[3],
  'EB_NoDox_D3_Flk1pos' = EB_cols[3],
  
  'EB_Dox_D0' = EB_cols[1],
  'EB_Dox_D25' = EB_cols[2],
  'EB_Dox_D25_Flk1pos' = EB_cols[2],
  'EB_Dox_D3' = EB_cols[3],
  'EB_Dox_D3_Flk1pos' = EB_cols[3]
)
is_flk1pos <- grepl('Flk1pos', colData(se)$group)
is_dox <- grepl('_Dox_', colData(se)$group)
pc_x <- 1; pc_y <- 2
plot(s$rotation[, pc_x], s$rotation[, pc_y], pch = 21, bg = group2col[colData(se)$group], col = group2col[colData(se)$group], cex = 2, xlab = sprintf('PC%d(%.3f%%)', pc_x, eigs[pc_x] / sum(eigs) * 100), ylab = sprintf('PC%d(%.3f%%)', pc_y, eigs[pc_y] / sum(eigs) * 100))
points(s$rotation[is_dox, pc_x], s$rotation[is_dox, pc_y], pch = 21, col = 'blue', cex = 2, lwd = 2)
points(s$rotation[is_flk1pos, pc_x], s$rotation[is_flk1pos, pc_y], pch = 3, col = 'pink', cex = 2, lwd = 2)
```
```{r, fig.width = 9, fig.height =6}
Z <- assays(dev)$z
rownames(Z) <- rowData(dev)$name
Z <- Z[order(v$p_value), ]
m <- v$p_value_adj < 0.05
Z <- Z[m, ]
Z <- Z[!duplicated(rownames(Z)), ]

column_annotation <- HeatmapAnnotation(
  stage = colData(se)$group, 
  flk1pos = is_flk1pos,
  dox = is_dox,
  col = list(stage = group2col, dox = c('TRUE' = 'blue', 'FALSE' = 'white'), flk1pos = c('TRUE' = 'pink', 'FALSE' = 'white'))
)

gs <- unique(c('AP-1(bZIP)', 'Brachyury(T-box)', 'CTCF(Zf)', 'Egr1(Zf)', 'Eomes(T-box)', 'Etv2(ETS)', 'FOXK2(Forkhead)', 'Gata2(Zf)', 'Oct4(POU,Homeobox)', 'RUNX1(Runt)', 'YY1(Zf)', 'Mef2c(MADS)', 'NRF1(NRF)', 'c-Myc(bHLH)', 'Isl1(Homeobox)', 'Klf4(Zf)', 'Brn1(POU,Homeobox)', 'Gata1(Zf)', 'Ets1-distal(ETS)', 'Fra1(bZIP)'))
mark_col <- rep('red', length(gs))
names(mark_col) <- gs

mark_at <- which(rownames(Z) %in% gs)
row_annotation <- rowAnnotation(
  up_genes = anno_mark(
    at = mark_at, 
    labels = rownames(Z)[mark_at], 
    labels_gp = gpar(fontsize = 15),
    padding = unit(1, "mm"),
    link_width = unit(10, "mm")
  ),
  simple_anno_size = unit(0.75, "cm")
)
col_fun <- colorRamp2(quantile(Z, c(0.05, 0.5, 0.95)), c("blue", "black", "yellow"))
Heatmap(Z, 
        cluster_rows = TRUE, 
        cluster_columns = TRUE, 
        column_split = colData(se)$cell,
        show_row_names = FALSE, 
        col = col_fun, 
        show_row_dend = TRUE,
        show_column_dend = FALSE,
        top_annotation = column_annotation,
        right_annotation = row_annotation
)
```


```{r}
sessionInfo()
```