---
title: "Comparing the Etv2 peaks at MEF D1 and EB 3h"
author: "Wuming Gong"
date: "12/31/2019"
output: rmarkdown::github_document
---

```{r, message = FALSE}
library(GenomicRanges)
library(ChIPpeakAnno)
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(clusterProfiler)
```

```{r}
maxgap_summit_overlap <- 200L
```

# Read the Etv2 ChIP-seq peaks in MEF (D1 post Etv2 induction) and D2.5 EB (3h post induction). 
```{r}
bed_files <- c(
  'MEF_Dox_d1_Etv2' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_Etv2_summits.bed',
  'EB_Dox_3h_Etv2' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_3h_Etv2_summits.bed'
)
gr_list <- lapply(1:length(bed_files), function(i){
  x <-read.table(bed_files[i], header = FALSE, sep = '\t')
  gr <- GRanges(seqnames = x[, 1], range = IRanges(x[, 2], x[, 3]), score = x[, 5])
  names(gr) <- x[, 4]
  gr
})
names(gr_list) <- names(bed_files)
```

# Looking at the overlap of Etv2 peaks
```{r}
ol <- findOverlapsOfPeaks(gr_list$MEF_Dox_d1_Etv2, gr_list$EB_Dox_3h_Etv2, maxgap = maxgap_summit_overlap)
names(ol$peaklist) <- c('EB', 'MEF', 'MEF&EB')
```

# Make a venn diagram of the overlapped peaks
Note that this chunk will fail within Rstudio due to the venneuler/rJava problem (e.g. https://github.com/s-u/rJava/issues/151)
There is no problem of executing this chunk in standalone R session.
```{r, eval = FALSE}
library(venneuler)
s <- sapply(ol$peaklist, length)
vd <- venneuler(s)
vd$labels <- c('', '')
plot(vd)
```


```{r}
anno_list <- lapply(ol$peaklist, function(p)
  annotatePeak(p, tssRegion = c(-2000, 500), TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, annoDb = "org.Mm.eg.db")
)
```

```{r}
plotAnnoBar(anno_list)
```

```{r}
genes <- lapply(anno_list, function(i) as.data.frame(i)$geneId)
genes <- lapply(genes, unique)
compKEGG <- compareCluster(geneCluster   = genes,
                         fun           = "groupGO",
                         OrgDb = 'org.Mm.eg.db',
                         ont = 'BP',
                         level = 3
                         )
```

```{r, fig.width = 10}
dotplot(compKEGG, showCategory = 30, title = "KEGG Pathway Enrichment Analysis")
```