---
title: 'Compare Etv2 ChIP-seq between D1 and D7 during MEF reprogramming
output: rmarkdown::github_document
date: '01/21/2019'
author: 'Wuming Gong'
---

This analysis identify the subgroups of Etv2 ChIP-seq peaks (D1 post induction in MEF) associated
with different MNase-seq levels and visualize them by heatmaps. 

```{r, include = FALSE}
devtools::load_all('packages/compbio')
library(dplyr)
library(grid)
library(BSgenome.Mmusculus.UCSC.mm10)
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(GO.db)
gr_url <- 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/all_Etv2_peaks.rds' 
```

### split the Etv2 MEF D1 peaks into three groups: 
1: on at D1 and off at D7 (early)
2: off at D1 and on at D7 (late)
3: on at both D1 and D7 (persist)
```{r}
gr <- readRDS(gzcon(url(gr_url)))
peaks <- gr[gr$group[, 'MEF_Dox_d1_Etv2'] | gr$group[, 'MEF_Dox_d7_Etv2']]
```

### Produce the normalized matrix for ComplexHeatmap plot and generate the heatmap
```{r}
gs <- 'Etv2_peaks_MEF_D1_Dox_merged'; extend <- 1000; w <- 50; smooth <- FALSE; target_ratio <- 0.2; mc.cores <- 2
bw_files <- c(
	'MEF_Dox_D1_Etv2' 									= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_Etv2_FE.bw',
	'MEF_Dox_D2_Etv2' 									= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d2_Etv2_FE.bw',
	'MEF_Dox_D7_Etv2' 									= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_Etv2_FE.bw',
	'MEF_NoDox_Brg1' 										= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_Brg1_FE.bw',
	'MEF_Dox_D1_Brg1' 									= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_Brg1_FE.bw',
	'MEF_Dox_D2_Brg1' 									= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d2_Brg1_FE.bw',
	'MEF_Dox_D7_Brg1' 									= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_Brg1_FE.bw',
  'MEF_NoDox_H3K27ac' 								= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw',
  'MEF_Dox_D1_H3K27ac' 								= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_H3K27ac_FE.bw',
  'MEF_Dox_D2_H3K27ac' 								= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d2_H3K27ac_FE.bw',
  'MEF_Dox_D7_H3K27ac' 								= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_H3K27ac_FE.bw',
	'MNase'                             = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/MNase_treat_pileup.bw',
	'H3'                                = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3_treat_pileup.bw',
	'H3K9me3'                           = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3K9me3_treat_pileup.bw',
	'H3K27me3'                          = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3K27me3_treat_pileup.bw',
	'H3K36me3'                          = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3K36me3_treat_pileup.bw',
	'H3K9ac'                            = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3K9ac_treat_pileup.bw',
	'H3K79me2'                          = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3K79me2_treat_pileup.bw',
	'H3K4me2'                           = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3K4me2_treat_pileup.bw',
	'H3K4me1'                           = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3K4me1_treat_pileup.bw',
	'Hdac1'                             = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/Hdac1_treat_pileup.bw',
	'H3.3'                              = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/H3.3_treat_pileup.bw',
	'P300'                              = '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/P300_treat_pileup.bw'
)
n2m_files <- peaks %>% 
	resize(fix = 'center', width = 1) %>% 
	get_normalizeToMatrix(gs, bw_files, extend = extend, w = w, smooth = smooth, target_ratio = target_ratio, mc.cores = mc.cores, force = FALSE)

group_cols <- rep('blue', length(bw_files)); names(group_cols) <- names(bw_files)
group_cols[c('H3', 'MNase')] <- 'green'
group_cols[c('MEF_Dox_D1_Etv2', 'MEF_Dox_D2_Etv2', 'MEF_Dox_D7_Etv2')] <- 'red'
group_cols[c('MEF_NoDox_Brg1', 'MEF_Dox_D1_Brg1', 'MEF_Dox_D2_Brg1', 'MEF_Dox_D7_Brg1')] <- 'purple'
group_cols[c('MEF_NoDox_H3K27ac', 'MEF_Dox_D1_H3K27ac', 'MEF_Dox_D2_H3K27ac', 'MEF_Dox_D7_H3K27ac')] <- 'darkgreen'
mat <- lapply(n2m_files, readRDS); names(mat) <- names(n2m_files)
col_fun <- lapply(1:length(mat), function(i) colorRamp2(quantile(mat[[i]], c(0.005, 0.995)), c('white', group_cols[i])))
names(col_fun) <- names(n2m_files)

#i <- 1:length(peaks)
set.seed(1); i <- sample(1:length(peaks), 2000)
sp <- factor(sprintf('%s_%s', peaks$group[, 'MEF_Dox_d1_Etv2'], peaks$group[, 'MEF_Dox_d7_Etv2']))
ta <- HeatmapAnnotation(enriched = anno_enriched(gp = gpar(lty = 1, lwd = 2, col = 1:nlevels(sp)), axis_param = list(facing = 'inside', at = -1000)))
h <- EnrichedHeatmap(mat[['MEF_Dox_D1_Etv2']][i, ], col = col_fun[['MEF_Dox_D1_Etv2']], split = sp[i], name = 'MEF_Dox_D1_Etv2', top_annotation = ta, pos_line = FALSE)

ss <- c('MEF_Dox_D2_Etv2', 'MEF_Dox_D7_Etv2', 'MEF_NoDox_Brg1', 'MEF_Dox_D1_Brg1', 'MEF_Dox_D2_Brg1', 'MEF_Dox_D7_Brg1', 'MEF_NoDox_H3K27ac', 'MEF_Dox_D1_H3K27ac', 'MEF_Dox_D2_H3K27ac', 'MEF_Dox_D7_H3K27ac')
for (s in ss[ss %in% names(bw_files)]){
  h <- h + EnrichedHeatmap(mat[[s]][i, ], col = col_fun[[s]], name = s, top_annotation = ta, pos_line = FALSE)
}
draw(h, heatmap_legend_side = 'bottom')
```

### split the Etv2 EB peaks into three groups: 
1: on at 3h and off at 12h (early)
2: off at 3h and on at 12h (late)
3: on at both 3h and 12h (persist)
```{r}
peaks <- gr[gr$group[, 'EB_Dox_3h_Etv2'] | gr$group[, 'EB_Dox_12h_Etv2']]

gs <- 'Etv2_peaks_EB_Dox_merged'; extend <- 1000; w <- 50; smooth <- FALSE; target_ratio <- 0.2; mc.cores <- 2
bw_files <- c(
	'EB_Dox_3h_Etv2' 					= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_3h_Etv2_FE.bw',
	'EB_Dox_12h_Etv2' 					= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_Etv2_FE.bw',
	'EB_Dox_3h_Brg1' =  '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_3h_Brg1_FE.bw',
	'EB_Dox_12h_Brg1' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_Brg1_FE.bw',
	'EB_NoDox_3h_Brg1' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_3h_Brg1_FE.bw',
	'EB_NoDox_12h_Brg1' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_Brg1_FE.bw',
	'EB_Dox_3h_H3K27ac' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_3h_H3K27ac_FE.bw',
	'EB_Dox_12h_H3K27ac' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_H3K27ac_FE.bw',
	'EB_NoDox_3h_H3K27ac' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_3h_H3K27ac_FE.bw',
	'EB_NoDox_12h_H3K27ac' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_H3K27ac_FE.bw'
)
n2m_files <- peaks %>%
  resize(fix = 'center', width = 1) %>%
  get_normalizeToMatrix(gs, bw_files, extend = extend, w = w, smooth = smooth, target_ratio = target_ratio, mc.cores = mc.cores, force = FALSE)

group_cols <- rep('blue', length(bw_files)); names(group_cols) <- names(bw_files)
group_cols[c('EB_Dox_3h_Etv2', 'EB_Dox_12h_Etv2')] <- 'red'
group_cols[c('EB_Dox_3h_Brg1', 'EB_Dox_12h_Brg1', 'EB_NoDox_3h_Brg1', 'EB_NoDox_12h_Brg1')] <- 'purple'
group_cols[c('EB_Dox_3h_H3K27ac', 'EB_Dox_12h_H3K27ac', 'EB_NoDox_3h_H3K27ac', 'EB_NoDox_12h_H3K27ac')] <- 'darkgreen'
mat <- lapply(n2m_files, readRDS); names(mat) <- names(n2m_files)
col_fun <- lapply(1:length(mat), function(i) colorRamp2(quantile(mat[[i]], c(0.005, 0.995)), c('white', group_cols[i])))
names(col_fun) <- names(n2m_files)

i <- 1:length(peaks)
#set.seed(1); i <- sample(1:length(peaks), 2000)
sp <- factor(sprintf('%s_%s', peaks$group[, 'EB_Dox_3h_Etv2'], peaks$group[, 'EB_Dox_12h_Etv2']))

ta <- HeatmapAnnotation(enriched = anno_enriched(gp = gpar(lty = 1, lwd = 2, col = 1:nlevels(sp)), axis_param = list(facing = 'inside', at = -1000)))
h <- EnrichedHeatmap(mat[['EB_Dox_3h_Etv2']][i, ], col = col_fun[['EB_Dox_3h_Etv2']], split = sp[i], name = 'EB_Dox_3h_Etv2', top_annotation = ta, pos_line = FALSE)
ss <- c('EB_Dox_12h_Etv2', 'EB_NoDox_3h_Brg1', 'EB_Dox_3h_Brg1', 'EB_Dox_12h_Brg1', 'EB_NoDox_3h_H3K27ac', 'EB_Dox_3h_H3K27ac', 'EB_Dox_12h_H3K27ac')
for (s in ss[ss %in% names(bw_files)]){
  h <- h + EnrichedHeatmap(mat[[s]][i, ], col = col_fun[[s]], name = s, top_annotation = ta, pos_line = FALSE)
}
draw(h, heatmap_legend_side = 'bottom')
```


## Look at the % of endothelial and neural genes near the late/early/persistn Etv2 peaks in both EB and MEF
```{r}
anno <- annotatePeak(gr, tssRegion = c(-5000, 1000), TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, annoDb = 'org.Mm.eg.db')
anno <- as.data.frame(anno)
groups <- list(	# six groups of genes 
	EB_late = unique(anno$geneId[!anno$group.EB_Dox_3h_Etv2 & anno$group.EB_Dox_12h_Etv2]),
	EB_early = unique(anno$geneId[anno$group.EB_Dox_3h_Etv2 & !anno$group.EB_Dox_12h_Etv2]),
	EB_persist= unique(anno$geneId[anno$group.EB_Dox_3h_Etv2 & anno$group.EB_Dox_12h_Etv2]),
	MEF_late  = unique(anno$geneId[!anno$group.MEF_Dox_d1_Etv2 & anno$group.MEF_Dox_d7_Etv2]),
	MEF_early  = unique(anno$geneId[anno$group.MEF_Dox_d1_Etv2 & !anno$group.MEF_Dox_d7_Etv2]),
	MEF_persist = unique(anno$geneId[anno$group.MEF_Dox_d1_Etv2 & anno$group.MEF_Dox_d7_Etv2])
)

gos <- c(
	'endothelium development' = 'GO:0003158',
	'vasculature development' = 'GO:0001944',
	'nervous system development' = 'GO:0007399',
	'cardiovascular system development' =  'GO:0072358',
	'stem cell population maintenance' = 'GO:0019827',
	'heart development' = 'GO:0007507',
	'blood vessel development' = 'GO:0001568'
)

gs <- mget(gos, org.Mm.egGO2ALLEGS)
names(gs) <- names(gos)
x <- do.call('rbind', lapply(groups, function(genes) do.call('cbind', lapply(gs, function(go) sum(genes %in% go) / length(genes)))))
rownames(x) <- names(groups)
t(x)

```

# Make a venn diagram of the overlapped late peaks
Note that this chunk will fail within Rstudio due to the venneuler/rJava problem (e.g. https://github.com/s-u/rJava/issues/151)
There is no problem of executing this chunk in standalone R session.
```{r, eval = FALSE}
length(unique(anno$geneId[anno$group.EB_Dox_12h_Etv2 & anno$group.MEF_Dox_d7_Etv2]))
length(unique(anno$geneId[!anno$group.EB_Dox_12h_Etv2 & anno$group.MEF_Dox_d7_Etv2]))
length(unique(anno$geneId[anno$group.EB_Dox_12h_Etv2 & !anno$group.MEF_Dox_d7_Etv2]))

sum(gr$group[, 'EB_Dox_12h_Etv2'] & gr$group[, 'MEF_Dox_d7_Etv2'])
sum(!gr$group[, 'EB_Dox_12h_Etv2'] & gr$group[, 'MEF_Dox_d7_Etv2'])
sum(gr$group[, 'EB_Dox_12h_Etv2'] & !gr$group[, 'MEF_Dox_d7_Etv2'])
```

