---
title: 'Heatmap for epigenetic data surrounding Etv2 ChIP-seq peaks at D1 post induction
output: rmarkdown::github_document
date: '12/26/2019'
author: 'Wuming Gong'
---

This analysis identify the subgroups of Etv2 ChIP-seq peaks (D1 post induction in MEF) associated
with different MNase-seq levels and visualize them by heatmaps. 

```{r, message = FALSE}
devtools::load_all('packages/compbio')
library(dplyr)
library(grid)
library(BSgenome.Mmusculus.UCSC.mm10)
```

# Read the Etv2 ChIP-seq peaks at D1 MEF reprogramming
```{r}
bed_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_Etv2_summits.bed'
peaks <- macs2.read_summits(bed_file)
seqlevels(peaks) <- seqlevels(BSgenome.Mmusculus.UCSC.mm10)
seqlengths(peaks) <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
seqinfo(peaks) <- seqinfo(BSgenome.Mmusculus.UCSC.mm10)
```

# Split the BAM file of MEF_NoDox ATAC-seq into mono nucleosome BAM and NFR BAM
```{bash}
INPUT_FILE=/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam
OUTPUT_NFR_FILE=/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_nfr.bam
OUTPUT_MONO_FILE=/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_mono_nucleosome.bam
samtools view -h $INPUT_FILE | awk '{if ($9 <= 100 && $9 >= -100 && $9 != 0 || $1 ~ /^@/) print $0}' | samtools view -bS - > $OUTPUT_NFR_FILE
samtools index $OUTPUT_NFR_FILE
samtools view -h $INPUT_FILE | awk '{if ((($9 <= 247 && $9 >= 180) || ($9 <= -180 && $9 >= -247)) && $9 != 0 || $1 ~ /^@/) print $0}' | samtools view -bS - > $OUTPUT_MONO_FILE
samtools index $OUTPUT_MONO_FILE
```

# Running MACS2 on the NFR and mono nucleosome BAM to generate mono / NFR ratio BW file for the heatmap
```{r}
dataset <- 'dataset=Etv2ATAC_version=20190228b'; genome <- 'mm10'
bname <- 'MEF_NoDox_mono_nucleosome_vs_nfr'
base.name <- sprintf('%s/%s', dataset_dir(dataset), bname)
treatment_files <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_mono_nucleosome.bam'
control_files <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_nfr.bam'
macs2.callpeak(treatment_files, base.name, control_files, format = 'BAMPE', genome = genome, broad = FALSE, qvalue.cutoff = 0.05, fold.change = TRUE, update = TRUE, call.summits = TRUE, shift = -100, extsize = 200)
```

# Produce the normalized matrix for ComplexHeatmap plot
```{r}
gs <- 'Etv2_peaks_MEF_D1_Dox_all'; extend <- 1000; w <- 50; smooth <- FALSE; target_ratio <- 0.2; mc.cores <- 4
bw_files <- c(
	'MNase' 														=	'/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/MNase_treat_pileup.bw',
	'MEF_NoDox_mono_nucleosome_vs_nfr' 	=	'/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_mono_nucleosome_vs_nfr_FE.bw',
  'MEF_Dox_D1_Etv2' 									= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_Etv2_FE.bw',
  'MEF_NoDox_Brg1' 										= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_Brg1_FE.bw',
  'MEF_NoDox_H3K27ac' 								= '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw'
)
n2m_files <- peaks %>% 
	resize(fix = 'center', width = 1) %>% 
	get_normalizeToMatrix(gs, bw_files, extend = extend, w = w, smooth = smooth, target_ratio = target_ratio, mc.cores = mc.cores, force = FALSE)

group_cols <- rep('blue', length(bw_files)); names(group_cols) <- names(bw_files)
group_cols['MNase'] <- 'green'
group_cols['MEF_Dox_D1_Etv2'] <- 'red'
group_cols['MEF_NoDox_Brg1'] <- 'purple'
group_cols['MEF_NoDox_H3K27ac'] <- 'darkgreen'
mat <- lapply(n2m_files, readRDS); names(mat) <- names(n2m_files)
col_fun <- lapply(1:length(mat), function(i) colorRamp2(quantile(mat[[i]], c(0.005, 0.995)), c('white', group_cols[i])))
names(col_fun) <- names(n2m_files)

#set.seed(1); i <- sample(length(peaks), 10000)
i <- seq_len(length(peaks))
mean_mnase <- rowMeans(mat[['MNase']][, (extend / w - 1):(extend / w  + 1)])
cls <- as.numeric(cut(mean_mnase, quantile(mean_mnase, c(1, 0.75, 0.25, 0)), include.lowest = TRUE))
sp <- factor(cls)
ta <- HeatmapAnnotation(enriched = anno_enriched(gp = gpar(lty = 1, lwd = 2, col = c(1, 3, 2)), axis_param = list(facing = 'inside', at = -1000)))
h <- EnrichedHeatmap(mat[['MNase']][i, ], col = col_fun[['MNase']], split = sp[i], name = 'MNase', top_annotation = ta, pos_line = FALSE)
ss <- c('MEF_NoDox_Brg1', 'MEF_NoDox_H3K27ac', 'MEF_Dox_D1_Etv2')
for (s in ss[ss %in% names(bw_files)]){
  h <- h + EnrichedHeatmap(mat[[s]][i, ], col = col_fun[[s]], name = s, top_annotation = ta, pos_line = FALSE)
}
draw(h, heatmap_legend_side = 'bottom')
```

# Save the mnase-seq intensity and clustering results
```{r}
peaks$mean_mnase <- mean_mnase
peaks$cluster <- cls
peak_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_Etv2_summits.rds'
saveRDS(peaks, peak_file)
s3_sync('datasets/dataset=Etv2PioneerChIPseq_version=20191203a/', make_public = TRUE)
```
