---
title: 'Preprocessing of scRNA-seq'
author: 
- name: 'Wuming Gong'
created: '12/05/2019'
updated: '12/05/2019'
output: rmarkdown::github_document
---

This document covers the preprocessing steps of the scRNA-seq of Etv2 reprogramming in MEF
The script for analyzing the resulting scRNA-seq data can be found at scRNA_seq_analysis.Rmd

```{r}
devtools::load_all('packages/compbio')
library(SummarizedExperiments)
library(igraph)
library(FNN)
library(slingshot)
library(RColorBrewer)
library(umap)
library(parallel)
dataset <- 'dataset=Etv2scRNAseq_version=20190416a'
nfeature <- 4000 # number HVG's
latent_dim <- 40	# number of latent dimension for scVI
k <- 50	# number of KNN neighbors for constructing KNN graph
```

Preprocessing the Etv2 scRNA-seq data
* Remove the potential doublets using scrublet
* Get the scaled and normalized gene expression matrix using Seurat 3.1
* Get the HVG's 

```{r, eval = FALSE}
se <- readRDS(sprintf('%s/filtered.rds', dataset_dir(dataset)))
se <- se[rowSums(assays(se)$counts > 0) >= 3, ]
se <- scrublet(se, colData(se)$group, expected_doublet_rate = 0.1, min_counts = 2, min_cells = 3, min_gene_variability_pctl = 85, n_prin_comps = 30L)
is_doublet <- colData(se)$doublet_scores > 0.25
se <- se[, !is_doublet]
se2 <- seurat_preprocess(se, nfeatures = nfeature)

# save the results
se_file <- sprintf('%s/filtered_nfeatures=%d.rds', dataset_dir(dataset), nfeature)
saveRDS(se2, se_file)
```

Using scVI for dimension reduction of scRNA-seq followed by UMAP visualization
This step takes ~1 hour

```{r, eval = FALSE}
se_file <- sprintf('%s/filtered_nfeatures=%d.rds', dataset_dir(dataset), nfeature)
n <- rowData(se)$is_hvg
se2 <- scvi(se[n, ], n_latent = latent_dim)
colData(se)$latent <- colData(se2)$latent
set.seed(1); y <- umap(colData(se)$latent)$layout
colData(se)$umap <- y

# save the results
se_file <- sprintf('%s/filtered_nfeatures=%d_latent=%d.rds', dataset_dir(dataset), nfeature, latent_dim)
saveRDS(se, se_file)
```

Cell clustering using louvain clustering
```{r, eval = FALSE}
se_file <- sprintf('%s/filtered_nfeatures=%d_latent=%d.rds', dataset_dir(dataset), nfeature, latent_dim)
se <- readRDS(se_file)

group2bg <- c(
	'MEF_Dox_D1' = 'black',
	'MEF_NoDox' = 'blue',
	'MEF_Dox_D2' = 'purple',
	'MEF_Dox_D7a' = 'red',
	'MEF_Dox_D7b' = 'pink'
)
knn <- get.knn(colData(se)$umap, k = k)
knn <- data.frame(from = rep(1:nrow(knn$nn.index), k), to = as.vector(knn$nn.index), weight = 1 / (1 + knn$nn.dist^2))
g <- graph_from_data_frame(knn, directed = FALSE)
g <- simplify(g)
lc <- cluster_louvain(g)
clust <- as.numeric(as.factor(membership(lc)))
G_max <- max(unique(clust))
colData(se)$cluster <- clust

plot(colData(se)$umap, col = colorRampPalette(brewer.pal(11,'Spectral'))(G_max)[clust], pch = 16, asp = 1, xaxt = 'n', yaxt = 'n', main = 'cluster', xlab = '', ylab = '')
y_centers <- do.call('rbind', lapply(1:G_max, function(i) apply(colData(se)$umap[clust == i, ], 2, median)))
text(y_centers[, 1], y_centers[, 2], 1:G_max, cex = 1.25)

```

Generate a MSP for the Louvain clusters
```{r, eval = FALSE}
start.clus <- '4'
set.seed(1); lin <- getLineages(colData(se)$umap, colData(se)$cluster, start.clus = start.clus)
dev.new(width = 5, height = 5); par(mar = c(2, 2, 2, 2))
bg <- group2bg[colData(se)$group]
plot(colData(se)$umap, col = bg, asp = 1, pch = 16, xaxt = 'n', yaxt = 'n', main = 'Lineages', xlab = '', ylab = '', cex = 0.5)
lines(lin, lwd = 3, show.constraints = TRUE)
```

Generate smooth curves from the starting cluster to various ending cluster(s)
This step takes ~30 - 60 mins
```{r, eval = FALSE}
set.seed(1); crv <- getCurves(lin, extend = 'n')
metadata(se)$curves <- crv
se_file <- sprintf('%s/filtered_nfeatures=%d_latent=%d_k=%d_slingshot.rds', dataset_dir(dataset), nfeature, latent_dim, k)
saveRDS(se, se_file)
```

[2019-12-06] Find the DE genes in each Louvain clusters by using the t-test on the scaled gene expression levels
```{r, eval = FALSE}
se_file <- sprintf('%s/filtered_nfeatures=%d_latent=%d_k=%d_slingshot.rds', dataset_dir(dataset), nfeature, latent_dim, k)
se <- readRDS(se_file)
G_max <- max(unique(colData(se)$cluster))
X <- assays(se)$scaled_counts
d <- expand.grid(cluster = 1:G_max, gene  = 1:nrow(se))
clust <- colData(se)$cluster
pvalues <- unlist(mclapply(1:nrow(d), function(i) t.test(X[d[i, 'gene'], clust == d[i, 'cluster']], X[d[i, 'gene'], clust != d[i, 'cluster']], alternative = 'greater')$p.value, mc.cores = 8))
P <- t(matrix(pvalues, nrow = G_max, ncol = nrow(se)))
rowData(se)$clusters <- P
saveRDS(se, se_file)
```

Save the final data onto s3 for the downstream analysis
```{r, eval = FALSE}
colData(se)$group <- colData(se)$group %>% factor(c('MEF_Dox_D1', 'MEF_NoDox', 'MEF_Dox_D2', 'MEF_Dox_D7a', 'MEF_Dox_D7b'))
colData(se)$group <- revalue(colData(se)$group, c("MEF_Dox_D1"="MEF_NoDox", "MEF_NoDox"="MEF_Dox_D1"))
saveRDS(se, 'data/processed_Etv2_scRNAseq.rds')
system('s3cmd put data/processed_Etv2_scRNAseq.rds s3://gongx030/etv2_pioneer/data/')
system('s3cmd setacl -P s3://gongx030/etv2_pioneer/data/processed_Etv2_scRNAseq.rds')
```


