---
title: "Clustering of scRNA-seq of Etv2 reprogramming"
output:
  html_document: default
date: "12/10/2019"
---

```{r setup, include=FALSE, eval = TRUE}
knitr::opts_chunk$set(fig.width = 5, fig.height =5)
knitr::opts_chunk$set(dev = 'pdf')
```


```{r, message = FALSE}
library(SummarizedExperiment)
library(RColorBrewer)
library(knitr)
library(dplyr)
library(cluster)
library(parallel)
library(ggplot2)
library(goseq)
library(stringr)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
```
Read the processed scRNA-seq data (~2.1G)
```{r}
se <- readRDS(gzcon(url('https://s3.msi.umn.edu/garry_projects/etv2_pioneer/processed_Etv2_scRNAseq.rds')))
```
Determine the number of clusters by running k-means on the scVI latent space
```{r elbow, echo=TRUE, fig.height=5, fig.width=5}
wss <- (nrow(colData(se)$umap)-1) * sum(apply(colData(se)$umap, 2, var))
for (i in 2:15) wss[i] <- sum(kmeans(colData(se)$umap,centers = i)$withinss)
plot(1:15, wss, type="b", xlab="Number of Clusters", ylab="Within groups sum of squares")
n_cluster <- 7
set.seed(1); cls <- kmeans(colData(se)$umap, n_cluster)$cluster
cls <- as.numeric(factor(cls, c(1, 5, 4, 2, 3, 6, 7)))
set.seed(1); colData(se)$cluster <- cls
abline(v = n_cluster, col = 'blue', lty = 2)
```
Figure of the cell clusters
```{r, fig.width = 4, fig.height =4}
group2bg <- c(
   'MEF_Dox_D1' = 'black',
   'MEF_NoDox' = 'blue',
   'MEF_Dox_D2' = 'purple',
   'MEF_Dox_D7a' = 'red',
   'MEF_Dox_D7b' = 'pink'
 )
plot(colData(se)$umap, col = colorRampPalette(brewer.pal(11,'Spectral'))(n_cluster)[colData(se)$cluster], pch = 16, asp = 1, cex = 0.2, main = 'K-means clusters', xlab = '', ylab = '', xlim = c(-10, 10), ylim = c(-10, 10))
y_centers <- do.call('rbind', lapply(1:n_cluster, function(i) apply(colData(se)$umap[colData(se)$cluster == i, ], 2, median)))
text(y_centers[, 1], y_centers[, 2], 1:n_cluster, cex = 1)
```
Figure of cell sources (e.g. time points)
```{r, fig.width = 4, fig.height = 4}
bg <- group2bg[colData(se)$group]
plot(colData(se)$umap, col = bg, asp = 1, pch = 16,  main = 'Cell sources', xlab = '', ylab = '', cex = 0.2, cex.main = 1, cex.axis = 1, xlim = c(-10, 10), ylim = c(-10, 10))
```
Identify the up-regulated genes in each cluster by using t-test on the scaled gene expression levels. 
This step takes a few mins to finish.
```{r}
X <- assays(se)$scaled_counts
d <- expand.grid(cluster = 1:n_cluster, gene  = 1:nrow(se))
clust <- colData(se)$cluster
pvalues <- unlist(mclapply(1:nrow(d), function(i) wilcox.test(X[d[i, 'gene'], clust == d[i, 'cluster']], X[d[i, 'gene'], clust != d[i, 'cluster']], alternative = 'greater')$p.value, mc.cores = 4)) # takes ~1 mins
P <- t(matrix(pvalues, nrow = n_cluster, ncol = nrow(se)))
rownames(P) <- rowData(se)$name
```

Expresson plot of a set of genes
```{r fig.width = 8, fig.height = 5}
X <- assays(se)$normalized_counts
rownames(X) <- rowData(se)$name
gs <- c(
   'Etv2', 'Emcn', 'Kdr', 'Lmo2', 'Cdh5', 'Sox18'
)
d <- do.call('rbind', lapply(gs, function(g){
   x <- X[g, ]
   m <- which(x > 0.1)
   data.frame(gene = g, x = colData(se)$umap[m, 1], y = colData(se)$umap[m, 2], expression = x[m])
}))

b <- c(0, 3, 6)
ggplot(d, aes(x = x, y = y, color = expression)) + geom_point(size = 0.25, shape = 20) +
   scale_color_gradientn(limits = c(0, max(b)), colours=c("white", "red", "darkred"), breaks=b, labels=format(b)) + 
   theme(panel.background = element_rect(fill = 'white', colour = 'black')) + 
   xlab('') + ylab('') + 
   facet_wrap(~ gene, ncol = 3) +
   theme(strip.text.x = element_text(size = 15, face = 'bold')) +
   theme(axis.text = element_text(size = 15, face = 'bold')) +
   xlim(-10, 10) + ylim(-10, 10)
```

Pathway analysis of highly expressed genes in each cluster

Create a list of signifiant genes up-regulated in each cluster, fit the Probability Weighting Function (PWF) and find the 
significantly enriched GO terms. 
```{r, warning = FALSE, message = FALSE}
h <- 7 # Flk1+ cell cluster
genes <- as.integer(P[, h] < 1e-10)
names(genes) <- rownames(P)
genes <- genes[!duplicated(names(genes))]
pwf <- nullp(genes, "mm10", "geneSymbol", plot.fit = FALSE)
go_res <- goseq(pwf, "mm10","geneSymbol", test.cats = c("GO:BP"))
```

```{r fig.width = 6, fig.height = 5}
go_res %>% 
   filter(numInCat < 1000) %>%
   top_n(7, wt = -over_represented_pvalue) %>% 
   mutate(hitsPerc = numDEInCat * 100 / numInCat, term = str_wrap(term, 20)) %>% 
   ggplot(aes(x = hitsPerc, y = term, colour = over_represented_pvalue, size = numDEInCat)) +
   geom_point() +
   labs(title = sprintf('cluster %d', h), x="Hits (%)", y="", colour="p value", size="Count") +
   theme(axis.text = element_text(size = 15, face = 'bold')) + 
   theme(legend.text = element_text(size = 15, face = 'bold')) + 
   theme(legend.title = element_text(size = 15, face = 'bold'))
```
```{r}
sessionInfo()
```
