---
title: 'Preprocessing of scRNA-seq'
date: "12/05/2019"
output: html_document
---

Preprocessing the Etv2 scRNA-seq data
Remove the potential doublet using scrublet

```{r}
devtools::load_all('packages/compbio')
dataset <- 'dataset=Etv2scRNAseq_version=20190416a'
se <- readRDS(sprintf('%s/filtered.rds', dataset_dir(dataset)))
i <- colSums(assays(se)$counts) > 5000
se <- se[, i]
saveRDS(se, sprintf('%s/filtered_5000.rds', dataset_dir(dataset)))
i <- rowSums(assays(se)$counts > 0) >= 5
se <- se[i, ]
saveRDS(se, sprintf('%s/filtered_5000_fivegenes.rds', dataset_dir(dataset)))
devtools::load_all('packages/compbio'); se <- scrublet(se, colData(se)$group, expected_doublet_rate = 0.06, min_counts = 2, min_cells = 3, min_gene_variability_pctl = 85, n_prin_comps = 30L)
se <- se[, !colData(se)$predicted_doublets]
saveRDS(se, sprintf('%s/filtered_5000_fivegenes_noDoublets.rds', dataset_dir(dataset)))
```

[2019-11-18] Preprocessing scRNA-seq using Seurat (3.1.1)
```{r}
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets.rds'
se <- readRDS(se_file)
library(BiocParallel)
register(MulticoreParam(16))
devtools::load_all('packages/compbio'); se2 <- scran_preprocess(se)
rowData(se2) <- cbind(rowData(se), rowData(se2))
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran.rds'
saveRDS(se2, se_file)
```

[2019-04-17] scVI analysis of the scRNA-seq of Etv2 induction in mouse
Run this on k40
[2019-04-23] SCRAN gives less HVGs
[2019-11-18] Updated analysis
```{r}
library(SummarizedExperiment)
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran.rds'
se <- readRDS(se_file)
n <- rowData(se)$FDR < 0.05
latent <- 10
devtools::load_all('packages/compbio'); se2 <- scvi(se[n, ], n_latent = latent)
colData(se)$latent <- colData(se2)$latent
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran_scVI.rds'
saveRDS(se, se_file)
```


[2019-04-23] TSNE of the Etv2 scRNA-seq on MEF
TSNE runs on the latent space obtained by scVI
[2019-11-18] Updated visualization using umap
```{r}
library(SummarizedExperiment)
library(SingleCellExperiment)
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran_scVI.rds'
se <- readRDS(se_file)

library(umap); set.seed(1); y <- umap(colData(se)$latent)$layout
colData(se)$umap <- y
#library(Rtsne); set.seed(1); colData(se)$tsne <- Rtsne(colData(se)$latent)$Y
#library(DDRTree); dt <- DDRTree(t(colData(se)$umap), verbose = TRUE, sigma = 1e-3, maxIter = 5)
group2bg <- c(
	'MEF_Dox_D1' = 'black', 
	'MEF_NoDox' = 'blue', 
	'MEF_Dox_D2' = 'purple', 
	'MEF_Dox_D7a' = 'red', 
	'MEF_Dox_D7b' = 'pink'
)
dev.new(width = 5, height = 5); par(mar = c(2, 2, 2, 2))
bg <- group2bg[colData(se)$group]
#plot(colData(se)$umap, cex = 0.5, pch = 21, bg  = bg, col = bg, xaxt = 'n', yaxt = 'n', xlab = '', ylab = '')
plot(colData(se)$tsne, cex = 0.5, pch = 21, bg  = bg, col = bg, xaxt = 'n', yaxt = 'n', xlab = '', ylab = '')

se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran_scVI_umap.rds'
saveRDS(se, se_file)
```

[2019-04-28] KNN graph and Louvain clustering
[2019-11-18] Updated analysis based on umap results
```{r}
library(igraph)
library(FNN)
devtools::load_all('packages/compbio')
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran_scVI_umap.rds'
se <- readRDS(se_file)
k <- 250
knn <- get.knn(colData(se)$umap, k = k)
knn <- data.frame(from = rep(1:nrow(knn$nn.index), k), to = as.vector(knn$nn.index), weight = exp(-as.vector(knn$nn.dist^2)))
g <- graph_from_data_frame(knn, directed = FALSE)
g <- simplify(g)
lc <- cluster_louvain(g)
clust <- as.numeric(as.factor(membership(lc)))
G_max <- max(unique(clust))
colData(se)$cluster <- clust
dev.new(width = 5, height = 5); par(mar = c(2, 2, 2, 2))
library(RColorBrewer); plot(colData(se)$umap, col = colorRampPalette(brewer.pal(11,'Spectral'))(G_max)[clust], pch = 16, asp = 1, xaxt = 'n', yaxt = 'n', main = 'cluster', xlab = '', ylab = '')
y_centers <- do.call('rbind', lapply(1:G_max, function(i) apply(colData(se)$umap[clust == i, ], 2, median)))
text(y_centers[, 1], y_centers[, 2], 1:G_max, cex = 1.25)
```


[2019-04-25] Use slingshot to get the lineages (a MST on the Mclust clusters)
[2019-11-18] Updated analysis by using umap
```{r}
start.clus <- '5'
library(slingshot); set.seed(1); lin <- getLineages(colData(se)$umap, colData(se)$cluster, start.clus = start.clus)
dev.new(width = 5, height = 5); par(mar = c(2, 2, 2, 2))
plot(colData(se)$umap, col = bg, asp = 1, pch = 16, xaxt = 'n', yaxt = 'n', main = 'Lineages', xlab = '', ylab = '', cex = 0.5)
lines(lin, lwd = 3, show.constraints = TRUE)
```



# ----------------------------------------------------------------------------
# [2019-04-25] Get the smooth curve for each lineage
# This step takes roughly ~2 hours; the intermediate results are saved and re-used later
# [2019-04-30] Setting extend to 'n' is critical for stablize the curve, otherwise some 
# curve may go back to the start points and form a loop
# [2019-11-19] Updated analysis 
# ----------------------------------------------------------------------------
set.seed(1); crv <- getCurves(lin, extend = 'n')	# this step ~ 30 mins
metadata(se)$curves <- crv
dev.new(width = 5, height = 5); par(mar = c(2, 2, 2, 2))
plot(colData(se)$umap, col = bg, asp = 1, pch = 16, xaxt = 'n', yaxt = 'n', main = 'trajectory', xlab = '', ylab = '', cex = 0.5)
cols <- rep('gray', 6); cols[5] <- 'green'
lines(metadata(se)$curves, lwd = 3, show.constraints = TRUE, col = cols)
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran_scVI_umap_slingshot.rds'
saveRDS(se, se_file)

# ----------------------------------------------------------------------------
# [2019-04-28] Visualize gene expression levels by density plot
# ----------------------------------------------------------------------------
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran_scVI_umap_slingshot.rds'
se <- readRDS(se_file)
library(roxygen2); library(devtools); devtools::document('packages/denviz')
X <- assays(se)$logcounts; rownames(X) <- rowData(se)$name
devtools::load_all('packages/denviz'); denviz(X, colData(se)$umap, g = 'Etv2', grid_points = 20)



# ---------------------------------------------------------------------
# [2019-11-19] chromVAR analysis of TFs dynamics of Etv2 reprogramming ATAC-seq data
# ---------------------------------------------------------------------
bed_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/ATAC_peaks_Etv2_reprogramming.bed'
devtools::load_all('packages/compbio'); peaks <- macs2.read_summits(bed_file)

bw_files <- c(
	'MEF_NoDox' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_NoDox_treat_pileup.bw',
	'MEF_Dox_D1' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D1_treat_pileup.bw',
	'MEF_Dox_D2' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D2_treat_pileup.bw',
	'MEF_Dox_D7' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D7_treat_pileup.bw',
	'MEF_Dox_D7_Flk1pos' = '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228a/MEF_Dox_D7_Flk1pos_treat_pileup.bw'
)
X <- do.call('cbind', mclapply(bw_files, function(bw_file){
	flog.info(sprintf('reading %s', bw_file))
	ga <- rtracklayer::import(bw_file, format = 'BigWig', which = reduce(peaks))
	ga <- add.seqinfo(ga, 'mm10')
	cvg <- coverage(ga, weight = as.numeric(mcols(ga)$score))
	sum(cvg[peaks])
}, mc.cores = 2))


library(chromVARmotifs) # https://github.com/GreenleafLab/chromVARmotifs
library(chromVAR)
library(BSgenome.Mmusculus.UCSC.mm10)
library(motifmatchr)
library(BiocParallel)
library(SummarizedExperiment)
register(MulticoreParam(4)) # Use 8 cores
motif.set <- 'mouse_pwms_v2'
data(list = motif.set, package = 'chromVARmotifs')
se <- SummarizedExperiment(assays = SimpleList(counts = X), rowRanges = peaks)
se <- addGCBias(se, genome = BSgenome.Mmusculus.UCSC.mm10)
motif_ix <- matchMotifs(get(motif.set), peaks, genome = 'mm10')
dev <- computeDeviations(object = se, annotations = motif_ix)
v <- computeVariability(dev)



# ----------------------------------------------------------------------------
# [2019-11-19] Find the genes that are DE along the endothelial trajectory
# ----------------------------------------------------------------------------
endothelial <- 5
se_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2scRNAseq_version=20190416a/filtered_5000_fivegenes_noDoublets_scran_scVI_umap_slingshot.rds'
se <- readRDS(se_file)
X <- assays(se)$logcounts; rownames(X) <- rowData(se)$name
P <- slingPseudotime(crv)
n <- order(P[, endothelial])	# ordering cells based on their positions in lineage #1
n <- n[!is.na(P[n, endothelial])]
X2 <- X[, n]
X2 <- X2[rowSums(X2 > 0) > 100, ]

S <- do.call('rbind', (mclapply(1:nrow(X2), function(g){
	x <- X2[g, ]
	df <- data.frame(x = 1:length(n), y = x)
	coef(summary(lm(y ~ x, df)))[2, ]
}, mc.cores = 16)))

padj <- p.adjust(S[, 4])


# -----------
is_dynamic_gene <- padj < 0.05 & S[, 1] > 0
is_dynamic_tf <- v$p_value_adj < 0.05 & assays(dev)$z[, 5] > assays(dev)$z[, 1]
Z <- assays(dev)$z
rownames(Z) <- rowData(dev)$name
Z <- Z[is_dynamic_tf, ]
Y <- X2
Y <- Y[is_dynamic_gene, ]
gs <- intersect(rownames(Z), rownames(Y))

library(ComplexHeatmap)
group2bg <- c(
	'MEF_Dox_D1' = 'black',
	'MEF_NoDox' = 'blue',
	'MEF_Dox_D2' = 'purple',
	'MEF_Dox_D7a' = 'red',
	'MEF_Dox_D7b' = 'pink'
)
ha <- HeatmapAnnotation(time = c('MEF_Dox_D1', 'MEF_NoDox', 'MEF_Dox_D2', 'MEF_Dox_D7a', 'MEF_Dox_D7b'), col = list(time = group2bg), show_annotation_name = FALSE)
Heatmap(Z[gs, ], col = colorRamp2(quantile(Z[gs, ], c(0.05, 0.5, 0.95)),  c("black", "darkgreen", "yellow")), cluster_columns = FALSE, cluster_rows = FALSE, show_column_names = FALSE, top_annotation = ha, row_names_gp = gpar(fontsize = 15))


Y2 <- t(apply(Y, 1, function(y) (y - min(y)) / (max(y) - min(y))))
ha <- HeatmapAnnotation(time = colData(se)$group[n], col = list(time = group2bg), show_annotation_name = FALSE)
Heatmap(as.matrix(Y2[gs, ]), col = colorRamp2(c(0.1, 0.5, 0.9),  c("black", "darkgreen", "yellow")), cluster_columns = FALSE, cluster_rows = FALSE, show_column_names = FALSE, top_annotation = ha, row_names_gp = gpar(fontsize = 15))



sc <- slingCurves(crv)[[endothelial]]

points(sc$s, lwd = 2, col = 'green')
g <- 'Emcn'
x <- X[g, ]
# get the cells and their order along the lineage #1
P <- slingPseudotime(crv)
n <- order(P[, endothelial])	# ordering cells based on their positions in lineage #1
n <- n[!is.na(P[n, endothelial])]
df <- data.frame(x = 1:length(n), y = x[n])
plot(y ~ x, df, pch = 21, bg = bg[n], col = bg[n], xlab = '', ylab = '', main = g)
#m <- df[, 'y'] > 0
m <- 1:nrow(df)
ss <- smooth.spline(df[m, 'x'], df[m, 'y'], df = 10)
lines(ss, lty = 2, col = 'black', lwd = 3)


# ----------------------------------------------------------------------------
# [2019-12-12] Preprocessing the Etv2 scRNA-seq data using Seurat 3.1
# ----------------------------------------------------------------------------
devtools::load_all('packages/compbio')
dataset <- 'dataset=Etv2scRNAseq_version=20190416a'
se <- readRDS(sprintf('%s/filtered.rds', dataset_dir(dataset)))
se <- se[rowSums(assays(se)$counts > 0) >= 3, ]
devtools::load_all('packages/compbio'); se <- scrublet(se, colData(se)$group, expected_doublet_rate = 0.1, min_counts = 2, min_cells = 3, min_gene_variability_pctl = 85, n_prin_comps = 30L)
is_doublet <- colData(se)$doublet_scores > 0.25
se <- se[, !is_doublet]

devtools::load_all('packages/compbio') 
ns <- c(1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000)
se_files <- sprintf('%s/filtered_nfeatures=%d.rds', dataset_dir(dataset), ns)
for (i in 1:length(ns)){
	se2 <- seurat_preprocess(se, nfeatures = ns[i])
	flog.info(sprintf('writing %s', se_files[i]))
	saveRDS(se2, se_files[i])
}



# ----------------------------------------------------------------------------
# [2019-12-03] scVI analysis of the scRNA-seq of Etv2 induction in mouse
# Run this on k40 (job 20191203a)
# ----------------------------------------------------------------------------



# ----------------------------------------------------------------------------
# [2019-12-03] Generate umap view of all n_features / latent_dim combinations
# ----------------------------------------------------------------------------
devtools::load_all('packages/compbio')
ns <- c(1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000); latent_dims <- c(10, 20, 30, 40, 50)
dataset <- 'dataset=Etv2scRNAseq_version=20190416a'
d <- expand.grid(nfeature = ns, latent_dim = latent_dims)
d <- transform(d, se_file = sprintf('%s/filtered_nfeatures=%d_latent=%d.rds', dataset_dir(dataset), nfeature, latent_dim))
d <- transform(d, pdf_file = sprintf('%s/filtered_nfeatures=%d_latent=%d.pdf', dataset_dir(dataset), nfeature, latent_dim))

library(umap)
group2bg <- c(
	 'MEF_Dox_D1' = 'black',
	 'MEF_NoDox' = 'blue',
	 'MEF_Dox_D2' = 'purple',
	 'MEF_Dox_D7a' = 'red',
	 'MEF_Dox_D7b' = 'pink'
)
res <- mclapply(1:nrow(d), function(i){
	flog.info(sprintf('processing %s', d[i, 'se_file']))
	se <- readRDS(d[i, 'se_file'])
	set.seed(1); y <- umap(colData(se)$latent)$layout
	colData(se)$umap <- y
#	bg <- group2bg[colData(se)$group]
#	pdf(d[i, 'pdf_file'], width = 10, height = 10)
#	plot(colData(se)$umap, cex = 0.5, pch = 21, bg  = bg, col = bg, xaxt = 'n', yaxt = 'n', xlab = '', ylab = '')
#	dev.off()
	saveRDS(se, d[i, 'se_file'])
}, mc.cores = 4)



# ----------------------------------------------------------------------------
# [2019-12-04] Generate the plot for clustering and SlingShot lineages for all the 
# combinations
# ----------------------------------------------------------------------------
devtools::load_all('packages/compbio')
#ns <- c(1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000); latent_dims <- c(10, 20, 30, 40, 50); ks <- c(50, 100, 150, 200, 250)
ns <- c(1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000); latent_dims <- c(60, 70, 80, 90, 100); ks <- c(50, 100, 150, 200, 250)
dataset <- 'dataset=Etv2scRNAseq_version=20190416a'
d <- expand.grid(nfeature = ns, latent_dim = latent_dims, k = ks)
d <- transform(d, se_file = sprintf('%s/filtered_nfeatures=%d_latent=%d.rds', dataset_dir(dataset), nfeature, latent_dim))
d <- transform(d, pdf_file = sprintf('%s/filtered_nfeatures=%d_latent=%d_k=%d.pdf', dataset_dir(dataset), nfeature, latent_dim, k))

library(igraph)
library(FNN)
library(slingshot)
group2bg <- c(
	 'MEF_Dox_D1' = 'black',
	 'MEF_NoDox' = 'blue',
	 'MEF_Dox_D2' = 'purple',
	 'MEF_Dox_D7a' = 'red',
	 'MEF_Dox_D7b' = 'pink'
)
res <- mclapply(1:nrow(d), function(i){
	flog.info(sprintf('processing %s', d[i, 'se_file']))
	se <- readRDS(d[i, 'se_file'])
	bg <- group2bg[colData(se)$group]
	knn <- get.knn(colData(se)$umap, k = d[i, 'k'])
	knn <- data.frame(from = rep(1:nrow(knn$nn.index),  d[i, 'k']), to = as.vector(knn$nn.index), weight = exp(-as.vector(knn$nn.dist^2)))
	g <- graph_from_data_frame(knn, directed = FALSE)
	g <- simplify(g)
	lc <- cluster_louvain(g)
	clust <- as.numeric(as.factor(membership(lc)))
	G_max <- max(unique(clust))
	flog.info(sprintf('writing %s', d[i, 'pdf_file']))
	pdf(d[i, 'pdf_file'], width = 10, height = 10)
	set.seed(1); lin <- getLineages(colData(se)$umap, clust)
	plot(colData(se)$umap, col = bg, asp = 1, pch = 16, xaxt = 'n', yaxt = 'n', main = 'Lineages', xlab = '', ylab = '', cex = 0.5)
	lines(lin, lwd = 3, show.constraints = TRUE)
	dev.off()
}, mc.cores = 4)



