---
title: 'Preparing the MEF ATAC-seq data for Etv2 peaks at D1 post induction
output: html_document
date: '12/26/2019'
author: 'Wuming Gong'
---

```{r, message = FALSE}
library(roxygen2)
library(devtools)
devtools::document('packages/seatac')
devtools::load_all('packages/compbio')
devtools::load_all('packages/seatac')
library(futile.logger); flog.threshold(TRACE)
library(BSgenome.Mmusculus.UCSC.mm10)
```


# Parameters
```{r}
peak_name <- 'MEF_Dox_d1_Etv2'; bam_name <- 'MEF_NoDox'
#window_size <- 320
window_size <- 640 
bin_size <- 5
fragment_size_range <- c(50, 370)
fragment_size_interval <- 5
latent_dim <- 10
#min_num_reads <- 10
min_num_reads <- 20
max_num_reads <- 500
genome <- 'mm10'
bed_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d1_Etv2_summits.bed'
mnase_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Chronis_version=20170519a/MNase_treat_pileup.bw'
bam_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam'
blacklist_file <- '/panfs/roc/scratch/gongx030/datasets/datasets=blacklists_version=20190827a/mm10.blacklist.bed.gz'
window_file <- sprintf('%s/windows/peak=%s_bam=%s_window_size=%d_bin_size=%d_fragment_size_range=%d+%d_fragment_size_interval=%d',
  project_dir('seatac'),
	peak_name,
  bam_name,
	window_size,
  bin_size,
	fragment_size_range[1],
	fragment_size_range[2],
	fragment_size_interval
)
model_dir_name <- sprintf('%s/models/peak=%s_bam=%s_latent_dim=%d_window_size=%d_bin_size=%d_fragment_size_range=%d+%d_fragment_size_interval=%d_min_num_reads=%d_max_num_reads=%d',
	project_dir('seatac'),
	peak_name,
	bam_name,
	latent_dim,
	window_size,
	bin_size,
	fragment_size_range[1],
	fragment_size_range[2],
	fragment_size_interval,
	min_num_reads,
	max_num_reads
)
```

# Prepare windows
```{r}
x <- read.table(bed_file, sep = '\t',  header = FALSE)
peaks <- GRanges(seqnames = x[, 1], range = IRanges(x[, 2], x[, 3]))
devtools::load_all('packages/seatac'); windows <- prepare_windows(
	peaks, 
	window_size = window_size, 
	bin_size = bin_size, 
	bigwig_files = c(MEF_MNase = mnase_file), 
	bam_files = bam_file, 
	blacklist_file = blacklist_file, 
	fragment_size_range = fragment_size_range, 
	fragment_size_interval = fragment_size_interval, 
	genome = BSgenome.Mmusculus.UCSC.mm10
)
saveRDS(windows, file = window_file)
```

# Running SeATAC
```{r}
windows <- readRDS(window_file)
windows <- windows[windows$num_reads >= min_num_reads & windows$num_reads <= max_num_reads]
set.seed(1); devtools::load_all('packages/seatac'); model <- seatac(windows, latent_dim = latent_dim, epochs = 100, batch_size = 128, steps_per_epoch = 50)
devtools::load_all('packages/seatac'); model %>% save_model(model_dir_name)
```

# Load the trained SeATAC model and make predictions
```{r}
devtools::load_all('packages/seatac'); gr <- model %>% predict(model$data, batch_size = 256)
predict_file <- sprintf('%s/predict.rds', model_dir_name)
saveRDS(gr, predict_file)
devtools::load_all('packages/compbio'); s3_sync('projects/seatac/', make_public = TRUE)
```


