---
title: 'Seatac on EB D2.5 ATAC-seq data surrounding EB 3h Etv2 ChIP-seq peak'
output: rmarkdown::github_document
date: '12/30/2019'
author: 'Wuming Gong'
---

```{r, message = FALSE}
library(roxygen2)
library(devtools)
devtools::document('packages/seatac')
devtools::load_all('packages/seatac')
devtools::load_all('packages/s3msi'); set_s3_environment()
library(futile.logger); flog.threshold(TRACE)
library(BSgenome.Mmusculus.UCSC.mm10)
```

# Parameters
```{r}
peak_name <- 'EB_Dox_3h_Etv2'; bam_name <- 'EB_NoDox_D25'
window_size <- 640 
bin_size <- 5
fragment_size_range <- c(50, 370)
fragment_size_interval <- 5
latent_dim <- 10
min_num_reads <- 20
max_num_reads <- 500
genome <- 'mm10'
bed_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_3h_Etv2_summits.bed'
mnase_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=mESC_MNase_version=20190925a/MNase_mESC.bw'
bam_file <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam'
blacklist_file <- '/panfs/roc/scratch/gongx030/datasets/datasets=blacklists_version=20190827a/mm10.blacklist.bed.gz'
window_file <- sprintf('%s/projects/seatac/windows/peak=%s_bam=%s_window_size=%d_bin_size=%d_fragment_size_range=%d+%d_fragment_size_interval=%d',
	Sys.getenv('TMPDIR'),
	peak_name,
  bam_name,
	window_size,
  bin_size,
	fragment_size_range[1],
	fragment_size_range[2],
	fragment_size_interval
)
model_dir_name <- sprintf('%s/projects/seatac/models/peak=%s_bam=%s_latent_dim=%d_window_size=%d_bin_size=%d_fragment_size_range=%d+%d_fragment_size_interval=%d_min_num_reads=%d_max_num_reads=%d',
	Sys.getenv('TMPDIR'),
	peak_name,
	bam_name,
	latent_dim,
	window_size,
	bin_size,
	fragment_size_range[1],
	fragment_size_range[2],
	fragment_size_interval,
	min_num_reads,
	max_num_reads
)
predict_file <- sprintf('%s/predict.rds', model_dir_name)
```

# Prepare windows
```{r}
x <- read.table(bed_file, sep = '\t',  header = FALSE)
peaks <- GRanges(seqnames = x[, 1], range = IRanges(x[, 2], x[, 3]))
devtools::load_all('packages/seatac'); windows <- prepare_windows(
	peaks, 
	window_size = window_size, 
	bin_size = bin_size, 
	bigwig_files = c(MEF_MNase = mnase_file), 
	bam_files = bam_file, 
	blacklist_file = blacklist_file, 
	fragment_size_range = fragment_size_range, 
	fragment_size_interval = fragment_size_interval, 
	genome = BSgenome.Mmusculus.UCSC.mm10
)
saveRDS(windows, file = window_file)
```

# Running SeATAC
```{r}
windows <- readRDS(window_file)
windows <- windows[windows$num_reads >= min_num_reads & windows$num_reads <= max_num_reads]
set.seed(1); devtools::load_all('packages/seatac'); model <- seatac(windows, latent_dim = latent_dim, epochs = 100, batch_size = 128, steps_per_epoch = 50)
devtools::load_all('packages/seatac'); model %>% save_model(model_dir_name)
```

# Load the trained SeATAC model and make predictions
```{r}
devtools::load_all('packages/seatac'); gr <- model %>% predict(model$data, batch_size = 256)
saveRDS(gr, predict_file)
s3_sync('projects/seatac/', make_public = TRUE)
```

```{r}
devtools::load_all('packages/s3msi'); local_to_public(predict_file)
devtools::load_all('packages/s3msi'); predict_file %>% s3_upload() %>% make_public()
```


# Split the BAM file of EB_NoDox_D25 ATAC-seq  into mono nucleosome BAM and NFR BAM
```{bash}
INPUT_FILE=/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam
OUTPUT_NFR_FILE=/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_nfr.bam
samtools view -h $INPUT_FILE | awk '{if ($9 <= 100 && $9 >= -100 && $9 != 0 || $1 ~ /^@/) print $0}' | samtools view -bS - > $OUTPUT_NFR_FILE
samtools index $OUTPUT_FILE

OUTPUT_MONO_FILE=/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_mono_nucleosome.bam
samtools view -h INPUT_FILE | awk '{if ((($9 <= 247 && $9 >= 180) || ($9 <= -180 && $9 >= -247)) && $9 != 0 || $1 ~ /^@/) print $0}' | samtools view -bS - > $OUTPUT_MONO_FILE
samtools index $OUTPUT_MONO_FILE
```


# Running MACS2 on the NFR and mono nucleosome BAM to generate NFR and mono specific BW files for the heatmap
```{r}
dataset <- 'dataset=Etv2ATAC_version=20190228b'; genome <- 'mm10'
bname <- 'EB_NoDox_D25_mono_nucleosome_vs_nfr'
base.name <- sprintf('%s/%s', dataset_dir(dataset), bname)
treatment_files <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_mono_nucleosome.bam'
control_files <- '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_nfr.bam'
macs2.callpeak(treatment_files, base.name, control_files, format = 'BAMPE', genome = genome, broad = FALSE, qvalue.cutoff = 0.05, fold.change = TRUE, update = TRUE, call.summits = TRUE, shift = -100, extsize = 200)
```

# Generate the normalize2Matrix files for the following heatmap
```{r}
gr <- readRDS(predict_file)

gs <- 'EB_Dox_3h_Etv2'; extend <- 1000; w <- 50; smooth <- FALSE; target_ratio <- 0.2; mc.cores <- 4
bw_files <- c(
  'MNase' =           '/panfs/roc/scratch/gongx030/datasets/dataset=mESC_MNase_version=20190925a/MNase_mESC.bw',
  'EB_NoDox_D25' =    '/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_treat_pileup.bw',
	'EB_Dox_3h_Etv2' = 	'/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_3h_Etv2_FE.bw',
	'EB_Dox_12h_Etv2' = 	'/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_Etv2_FE.bw',
	'EB_NoDox_D25_mono_vs_nfr' = 	'/panfs/roc/scratch/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_mono_nucleosome_vs_nfr_FE.bw',
	'EB_NoDox_3h_Brg1' = 	'/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_3h_Brg1_FE.bw',
	'EB_NoDox_3h_H3K27ac' = 	'/panfs/roc/scratch/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_3h_H3K27ac_FE.bw'
)
#devtools::load_all('packages/s3msi'); bw_files %>% local_to_s3() %>% s3_download()

devtools::load_all('packages/compbio'); n2m_files <- normalizeToMatrix_files(gs, base_names = names(bw_files), extend = extend, w = w, smooth = smooth, target_ratio = target_ratio)
#devtools::load_all('packages/s3msi'); n2m_files %>% local_to_s3() %>% s3_download()
devtools::load_all('packages/compbio'); get_normalizeToMatrix(bw_files, n2m_files, gr, extend = extend, w = w, smooth = smooth, target_ratio = target_ratio, mc.cores = mc.cores, force = FALSE)
```


